### PR TITLE
Handle dragging with NVDA/JAWS in browse mode

### PR DESCRIPTION
**Pull request by devongovett**
Related to #3221

After https://github.com/adobe/react-spectrum/pull/3421, it isn't possible to trigger drag and drop while in browse mode in NVDA/JAWS. This is the default when navigating to button elements, like in the default dnd story. In browse mode, only an onClick event is fired. Keyboard events are not fired.

Solution is to add an onClick handler that only applies if a virtual click is detected. This means that there can be no other click events on the element, because we cannot distinguish Enter from Space. For standalone draggable elements I think this is probably ok. For elements within collections, where we might have selection, we remove the onClick handler, which is ok because NVDA/JAWS are always in forms mode within collections.

This also centralizes the isVirtualClick and isVirtualPointerEvent functions into `@react-aria/utils`, and fixes dropping on button drop targets in browse mode.

### CODE REVIEW

**Review by majornista — APPROVED** [review]
This seems to be working well when I merge it into #3221.

File: packages/@react-aria/dnd/src/useDrag.ts [hunk 1]
```diff
@@ -213,17 +213,7 @@ export function useDrag(options: DragOptions): DragResult {
   };
 
   let modality = useDragModality();
-  let message: string;
-  if (!isDragging) {
-    if (modality === 'touch' && !hasDragButton) {
-      message = 'dragDescriptionLongPress';
```
Comment (snowystinger) [inline packages/@react-aria/dnd/src/useDrag.ts:219]: were we able to delete a string?

Comment (devongovett) [reply packages/@react-aria/dnd/src/useDrag.ts:219]: No, but now standalone drags don't need long press. That's only needed within a collection.

Comment (snowystinger) [reply packages/@react-aria/dnd/src/useDrag.ts:219]: ahhhh, thanks

**Review by snowystinger — DISMISSED** [review]
I'm unable to test in NVDA/JAWS, but the logic looks good

File: packages/@react-aria/dnd/src/useDrag.ts [hunk 2]
```diff
@@ -304,6 +297,14 @@ export function useDrag(options: DragOptions): DragResult {
           e.stopPropagation();
           startDragging(e.target as HTMLElement);
         }
+      },
+      onClick(e) {
+        // Handle NVDA/JAWS in browse mode, and touch screen readers. In this case, no keyboard events are fired.
+        if (isVirtualClick(e.nativeEvent) || modalityOnPointerDown.current === 'virtual') {
+          e.preventDefault();
+          e.stopPropagation();
+          startDragging(e.target as HTMLElement);
```
Comment (majornista) [inline packages/@react-aria/dnd/src/useDrag.ts:306]: If we're already dragging, clicking the drag target should cancel dragging. In this case, the drag operation will cancel and then immediately start dragging again.

**Review by majornista — CHANGES_REQUESTED** [review]

**Review by majornista — CHANGES_REQUESTED** [review]

File: packages/@react-aria/dnd/src/useDrag.ts [hunk 3]
```diff
@@ -304,6 +297,14 @@ export function useDrag(options: DragOptions): DragResult {
           e.stopPropagation();
           startDragging(e.target as HTMLElement);
         }
+      },
```
[the remark this answers is not in the document]
Comment (devongovett) [reply packages/@react-aria/dnd/src/useDrag.ts:300]: I think dropping on key up makes the most sense.

**Review by snowystinger — DISMISSED** [review]

**Review by LFDanLu — COMMENTED** [review]
Its the oddest thing, the built [storybook](https://reactspectrum.blob.core.windows.net/reactspectrum/3585d3bf4f764321ce34321841ed58315ecdde7a/storybook/index.html?path=/story/drag-and-drop--default&providerSwitcher-toastPosition=bottom) doesn't start drag when I hit Enter with NVDA, but it works when I have the branch checked out and use the locally hosted storybook on my Windows machine... Trying with a new build

**Review by LFDanLu — CHANGES_REQUESTED** [review]
Looks like the default DnD story doesn't work in NVDA if you open it outside the iframe. `modalityOnPointerDown` isn't being set to `virtual` in `useDrag`'s `onPointerDown` but this can be fixed by changing this line to `<=` checks instead:
https://github.com/adobe/react-spectrum/blob/8e954b87fa36c0b80b39f4d90961753c7dbcab01/packages/%40react-aria/dnd/src/useDrag.ts#L280

The `offsetY` and `centerY` values I got when testing outside the iframe were `20.5` and `20` respectively which just barely doesn't pass that if statement. Seems to work ok in Windows Chrome and FF after that

**Comment by devongovett** [comment]
This is weird because that line was really meant for Android Talkback but ok...

**Review by LFDanLu — APPROVED** [review]
Confirmed it works now w/ NVDA both inside and outside the iframe

**Review by dannify — APPROVED** [review]

**Pull request merged**